### PR TITLE
Migrate to fp-ts v2

### PR DESCRIPTION
**Pull request by gabro**

### CODE REVIEW

File: src/commands/add.ts [hunk 1]
```diff
@@ -28,26 +29,24 @@ export const builder: { [key: string]: yargs.Options } = {};
 
 export function handler(): Promise<unknown> {
   logInfo(":palm_tree: Getting a break? Great!\n");
-  return askWhere()
-    .chain(where =>
-      askWhen().chain(when => {
+  return mapLeft(
+    taskEither.chain(askWhere(), where =>
+      taskEither.chain(askWhen(), when => {
```
Comment (gabro) [inline src/commands/add.ts]: 😟

- business logic flow is lost
- we begin with `mapLeft`, which is then only "closed" at line 48

File: src/commands/add.ts [hunk 2]
```diff
@@ -60,43 +59,53 @@ function askPrompt(questions: Questions): TaskEither<unknown, Answers> {
...
 
 function askWhere(): TaskEither<unknown, string> {
-  return askPrompt([
-    {
-      type: "input",
-      name: "where",
-      message: "Where are you going?"
-    }
-  ]).map(a => a.where);
+  return taskEither.map(
```
Comment (gabro) [inline src/commands/add.ts]: 😕again, reversed business logic, where asking an input and __then__ mapping, but we're forced to write it in reverse order

File: src/commands/add.ts [hunk 3]
```diff
@@ -60,43 +59,53 @@ function askPrompt(questions: Questions): TaskEither<unknown, Answers> {
...
-    ])
-      .map(a => a.when_to)
-      .map(toDate => ({
+    ]),
+    a => a.when_from
+  );
+
+  return taskEither.chain(from, fromDate =>
+    taskEither.map(
+      taskEither.map(
```
Comment (gabro) [inline src/commands/add.ts]: 😟 chain, map, map... oh, finally our logic! And then a bunch of callbacks :(

File: src/commands/add.ts [hunk 4]
```diff
@@ -207,14 +210,12 @@ type Context = {
...
-  return manageExistingEvents().chain(createEventOnCalendar);
+  return readerTaskEither.chain(manageExistingEvents(), createEventOnCalendar);
 }
 
 function addOnTeamweek(): ReaderTaskEither<Context, unknown, void> {
   logInfo("\n:hourglass: Adding your vacation to Teamweek...\n");
-  return ask<Context, unknown>().chain(({ when, where }) =>
-    fromTaskEither(addTask(where, when))
-  );
+  return ({ when, where }) => addTask(where, when);
```
Comment (gabro) [inline src/commands/add.ts:237]: 😃readers are just functions! 🕺

File: src/teamweek.ts [hunk 5]
```diff
@@ -107,8 +114,8 @@ function authenticatedRequest<A, B = {}>(
   method: "GET" | "POST",
   body?: B
 ): TaskEither<unknown, A> {
-  return authenticate()
-    .chain(token =>
+  return mapLeft(
```
Comment (gabro) [inline src/teamweek.ts]: 😟 again, starting with `mapLeft`, which is a later conern

File: src/teamweek.ts [hunk 6]
```diff
@@ -170,34 +179,38 @@ export function getMe(): TaskEither<unknown, User> {
 }
 
 export function getBuildoWorkspace(): TaskEither<unknown, Workspace> {
-  return getMe().chain(user =>
-    array
-      .findFirst(user.workspaces, w => w.name.toLowerCase() === "buildo")
-      .fold(fromLeft("'buildo' workspace not found"), taskEither.of)
+  return taskEither.chain(getMe(), user =>
+    fold(
```
Comment (gabro) [inline src/teamweek.ts:184]: 🤔 some functions come from instance dictionaries, some functions are "free".
I know the technical reason, but the API feels arbitrary and I constantly wonder: "do I search for this in the instance or is it a free function?".

At the very beginning, I've used the free `chain` function, which obviously didn't compile. Then I've spent a good minute or two searching for `option.fold` which doesn't exist.

File: src/teamweek.ts [hunk 7]
```diff
@@ -170,34 +179,38 @@ export function getMe(): TaskEither<unknown, User> {
...
 export function addTask(
   name: string,
   interval: Interval
 ): TaskEither<unknown, void> {
-  return getMe().chain(user =>
-    getBuildoWorkspace().chain(workspace =>
-      getFerieProject().chain(project =>
+  return taskEither.chain(getMe(), user =>
+    taskEither.chain(getBuildoWorkspace(), workspace =>
+      taskEither.chain(getFerieProject(), project =>
```
Comment (gabro) [inline src/teamweek.ts]: 😟send HALP

File: src/commands/add.ts [hunk 8]
```diff
@@ -171,33 +182,25 @@ function manageEventsCreated(
...
-        const eventsCreated = response.items.filter(
-          event => event.creator.self
-        );
-        return manageEventsCreated(eventsCreated).chainSecond(
-          manageEventsInvited(eventsInvited)
-        );
-      })
-    )
-  );
+  return ({ when }) =>
```
Comment (gabro) [inline src/commands/add.ts:189]: 😃readers are just functions! 🕺

File: src/commands/add.ts [hunk 9]
```diff
@@ -171,33 +182,25 @@ function manageEventsCreated(
...
-      createEvent(when, `Ferie ${where}`).map(events => {
-        logInfo(
-          "\n:white_check_mark: Created event on your calendar. Click here to see it:\n"
-        );
-        logDetail(events.htmlLink);
-        logDetail("\n");
-      })
-    )
-  );
+  return ({ when, where }) =>
```
Comment (gabro) [inline src/commands/add.ts]: 😃readers are just functions! 🕺

File: src/commands/add.ts (hunk 1 above)
Comment (giogonzo) [reply src/commands/add.ts]: Have you also experimented with `Do` or `Fluent` for these longer business logic portions?

Comment (gabro) [reply src/commands/add.ts]: I'm about to

File: src/gcal.ts [hunk 10]
```diff
@@ -37,83 +43,104 @@ const calendar = google.calendar({
...
+  return fluent(taskEither)(taskify(oauth2Client.getToken)(code))
+    .chain(
+      credentials => saveCredentials(credentials, oauth2Client) as any // TODO
+    )
+    .mapLeft(err =>
+      fluent(taskEither)(
+        logError(`${err.response.status} ${err.response.statusText}`)
+      )
+        .chain(() => logError(JSON.stringify(err.response.data)))
+        .value()
```
Comment (gabro) [inline src/gcal.ts:76]: @gcanti sarebbe comodo avere qui un `chainLeft` di qualche tipo. `logError` torna un `TaskEither` e qui ho dovuto runnarlo a mano.

Comment (gcanti) [reply src/gcal.ts:76]: @gabro non capisco bene l'operazione, se `logError` torna un `TaskEither` quella `mapLeft` mi pare strana

Comment (gabro) [reply src/gcal.ts:76]: sostanzialmente voglio loggare in caso di errore, ma l'operazione di log torna un TaskEither

Comment (gcanti) [reply src/gcal.ts:76]: ok ma la firma come sarebbe?

```ts
declare function chainLeft<A, L, M, B>(ma: TaskEither<L, A>, f: (l: L) => TaskEither<M, B>): TaskEither<?, ?>
```

**Pull request closed without merging**